src/python/pants/jvm/compile.py

```diff
@@ -47,6 +54,34 @@ class _ClasspathEntryRequestClassification(Enum):
     INCOMPATIBLE = auto()
 
 
+@dataclass(frozen=True)
+class JVMRequestTypes:
+    classpath_entry_requests: tuple[type[ClasspathEntryRequest], ...]
+    code_generator_requests: FrozenDict[type[SourcesField], type[ClasspathEntryRequest]]
+
+
+@rule
+def calculate_jvm_request_types(union_membership: UnionMembership) -> JVMRequestTypes:
```

Comment:
This strategy looks great, and is likely generalizable to all of codegen (i.e., to be shared with other languages). But no need to worry about that here.